esp-radio/src/wifi/csi.rs

```diff
@@ -1,26 +1,341 @@
+        esp_wifi_set_csi_config,
+        esp_wifi_set_csi_rx_cb,
+        wifi_csi_config_t,
+        wifi_csi_info_t,
+    },
+    wifi::wifi_pkt_rx_ctrl_t,
 };
 
-pub(crate) trait CsiCallback: FnMut(wifi_csi_info_t) {}
+/// CSI (Channel State Information) packet metadata and associated packet details.
+///
+/// This structure contains the raw CSI data, along with necessary metadata
+/// from the received WiFi packet (MAC addresses, sequence number, packet headers).
+#[repr(transparent)]
+pub struct WifiCsiInfo<'a> {
+    inner: *const wifi_csi_info_t,
+    _lt: PhantomData<&'a ()>,
+}
+
+impl<'a> WifiCsiInfo<'_> {
+    /// Metadata header for the CSI data of a received packet.
+    pub fn rx_ctrl(&self) -> WifiPktRxCtrl<'a> {
+        WifiPktRxCtrl {
+            inner: unsafe { &(*self.inner).rx_ctrl },
+            _lt: PhantomData,
+        }
+    }
+
+    /// Source MAC address of the CSI data.
+    pub fn mac(&self) -> &[u8; 6] {
+        unsafe { &(*self.inner).mac }
+    }
+
+    /// Destination MAC address of the CSI data.
+    pub fn destination_mac(&self) -> &[u8; 6] {
+        unsafe { &(*self.inner).dmac }
+    }
+
+    /// First four bytes of the CSI data is invalid or not, true indicates the first four bytes is
+    /// invalid due to hardware limitation.
+    pub fn first_word_invalid(&self) -> bool {
+        unsafe { (*self.inner).first_word_invalid }
+    }
+
+    /// Valid buffer of CSI data.
+    pub fn buf(&self) -> &[i8] {
+        unsafe {
+            if (*self.inner).buf.is_null() || (*self.inner).len == 0 {
+                &[]
+            } else {
+                core::slice::from_raw_parts((*self.inner).buf, (*self.inner).len as usize)
+            }
+        }
+    }
+
+    /// Header of the wifi packet.
+    pub fn header(&self) -> &[u8] {
+        unsafe {
+            if (*self.inner).hdr.is_null() {
+                &[]
+            } else {
+                const HDR_LEN: usize = 24;
+                core::slice::from_raw_parts((*self.inner).hdr, HDR_LEN)
+            }
+        }
+    }
+
+    /// Payload of the wifi packet.
+    pub fn payload(&self) -> &[u8] {
+        unsafe {
+            if (*self.inner).payload.is_null() || (*self.inner).payload_len == 0 {
+                &[]
+            } else {
+                core::slice::from_raw_parts(
+                    (*self.inner).payload,
+                    (*self.inner).payload_len as usize,
+                )
+            }
+        }
+    }
+
+    /// Rx sequence number of the wifi packet.
+    pub fn rx_sequence(&self) -> u16 {
+        unsafe { (*self.inner).rx_seq }
+    }
+}
+
+/// Received packet radio metadata header.
+#[repr(transparent)]
+pub struct WifiPktRxCtrl<'a> {
+    inner: *const wifi_pkt_rx_ctrl_t,
+    _lt: PhantomData<&'a ()>,
+}
+#[cfg(not(esp32c6))]
+impl WifiPktRxCtrl<'_> {
+    /// Received Signal Strength Indicator (RSSI) of the packet.
+    pub fn rssi(&self) -> i8 {
+        unsafe { (*self.inner).rssi() as i8 }
```

Comment:
You removed this code from the example

```rust
// Signed bitfields are broken in rust-bingen, see https://github.com/esp-rs/esp-wifi-sys/issues/482
            let rssi = if rx_ctrl.rssi() > 127 {
                rx_ctrl.rssi() - 256
            } else {
                rx_ctrl.rssi()
            };
```

Which is good - but then we should fix the value here - I guess we should do that for other signed-integers as well